examples/bench/main.go

```diff
@@ -0,0 +1,1020 @@
+	px416 *ebiten.Image
+)
+
+var benchList = []benchmark{
+	{
+		name: "draw26",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px26, nil)
+			}
+		},
+	},
+	{
+		name: "draw104",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px104, nil)
+			}
+		},
+	},
+	{
+		name: "draw416",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			for i := 0; i < b.N; i++ {
+				screen.DrawImage(px416, nil)
+			}
+		},
+	},
+	{
+		name: "draw26color1",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			ops := []*ebiten.DrawImageOptions{&ebiten.DrawImageOptions{}, &ebiten.DrawImageOptions{}}
+			ops[0].ColorM.Scale(1.0, 0.2, 0.2, 1.0)
+			ops[1].ColorM.Scale(0.2, 1.0, 0.2, 1.0)
+			for i := 0; i < b.N; i++ {
+				idx := i
+				ops[idx%2].GeoM.Reset()
+				ops[idx%2].GeoM.Translate(float64(i*320/b.N), float64(i*200/b.N))
+				screen.DrawImage(px26, ops[idx%2])
+			}
+		},
+	},
+	{
+		name: "draw26color10",
+		fn: func(b *testing.B, screen *ebiten.Image) {
+			ops := []*ebiten.DrawImageOptions{&ebiten.DrawImageOptions{}, &ebiten.DrawImageOptions{}}
+			ops[0].ColorM.Scale(1.0, 0.2, 0.2, 1.0)
+			ops[1].ColorM.Scale(0.2, 1.0, 0.2, 1.0)
+			for i := 0; i < b.N; i++ {
+				idx := i / 10
+				ops[idx%2].GeoM.Reset()
+				ops[idx%2].GeoM.Translate(float64(i*320/b.N), float64(i*200/b.N))
+				screen.DrawImage(px26, ops[idx%2])
+			}
+		},
+	},
+}
+
+var (
+	benchIdx  = 0
+	results   []testing.BenchmarkResult
+	didSetup  sync.Once
+	benchFunc func(b *testing.B, screen *ebiten.Image)
+	benchGo   = make(chan struct{})
+	benchDone = make(chan bool)
+)
+
+func benchmarks(screen *ebiten.Image) error {
+	var setupErr error
+
+	didSetup.Do(func() {
+		eimg, _, err := openEbitenImage()
+		if err != nil {
+			setupErr = fmt.Errorf("can't open image: %s", err)
+			return
+		}
+		px26 = eimg
+
+		w, h := eimg.Size()
+		img, err := ebiten.NewImage(w*4, h*4, ebiten.FilterNearest)
+		if err != nil {
+			setupErr = fmt.Errorf("can't create ne image: %s", err)
+			return
+		}
+		px104 = img
+
+		op := &ebiten.DrawImageOptions{}
+		op.GeoM.Scale(4, 4)
+		px104.DrawImage(px26, op)
+
+		img, err = ebiten.NewImage(w*16, h*16, ebiten.FilterNearest)
+		if err != nil {
+			setupErr = fmt.Errorf("can't create ne image: %s", err)
+			return
+		}
+		px416 = img
+		op.GeoM.Scale(4, 4)
+		px416.DrawImage(px26, op)
+	})
+	if setupErr != nil {
+		return setupErr
+	}
+
+	if benchFunc != nil {
+		benchGo <- struct{}{}
+		if <-benchDone {
+			benchFunc = nil
+		}
+	} else {
+		if len(benchList) > benchIdx {
+			benchFunc = benchList[benchIdx].fn
+			go func(idx int) {
+				benchList[idx].result = testing.Benchmark(func(b *testing.B) {
+					b.StopTimer()
+					ebiten.SetBenchmark(b)
+					var elapsed time.Duration
+					// run for at least 60 frames, so the FPS value should mean something
+					for i := 0; i < 60; i++ {
+						<-benchGo
+						t0 := time.Now()
+						b.StartTimer()
```

Comment:
As `StartTimer` is called here, why does Ebiten side call `StartTimer`?

Reply:
The other side happens *after* the update. When the entire update is done, there's a later point outside the update when we hit ResolveStaleImages, and that is where the majority of the actual time is taken. So the idea is, I run the timer for the small window where I'm actually running the benchmark code, and then run it again for the GL code that does the actual rendering.

It might actually make more sense to just put the start/stop on the ebiten side exclusively, but put them around the entire update. The current form evolved from me initially just benchmarking the actual draw code in the user-provided update function, and then realizing that wasn't reflecting actual workload at all, because the huge expense of GL draw calls isn't happening during that code.

Reply:
OK now I think I understand what `StartTimer` and `StopTimer`, which are needed not to measure noisy things. My new question is, is it safe to call `StopTimer/StartTimer` pairs in different goroutines? Maybe I'm still not sure what you are trying to measure.

Reply:
So far as I know, it's safe as long as we know they're not concurrent, which should be the case; the call in the ebiten code is in the same path as the call to the user-provided update.

The goal is to measure "actual rendering time", roughly, which is the sum of the time we spend processing things like DrawImage calls, and the time spent feeding them to OpenGL. So the numbers reported by this version of the benchmark correlate fairly closely with how much impact a given drawing operation has on frame rate.